src/js/wrap.cpp

```diff
@@ -402,51 +384,170 @@ namespace ccf::js
     return JS_UNDEFINED;
   }
 
-  static bool _check_kv_map_access(
-    TxAccess access, const std::string& table_name)
+  enum class MapAccessPermissions
   {
-    const auto [security_domain, access_category] =
+    READ_WRITE,
+    READ_ONLY,
+    ILLEGAL
+  };
+
+  static constexpr char const* access_permissions_explanation_url =
+    "https://microsoft.github.io/CCF/main/audit/read_write_restrictions.html";
+
+  static MapAccessPermissions _check_kv_map_access(
+    TxAccess execution_context, const std::string& table_name)
+  {
+    // Enforce the restrictions described in the read_write_restrictions page in
+    // the docs
+    const auto [privacy_of_table, namespace_of_table] =
       kv::parse_map_name(table_name);
 
-    bool read_only = false;
-    switch (access_category)
+    switch (privacy_of_table)
     {
-      case kv::AccessCategory::INTERNAL:
+      case (kv::SecurityDomain::PRIVATE):
       {
-        if (security_domain == kv::SecurityDomain::PUBLIC)
+        switch (execution_context)
         {
-          read_only = true;
-        }
-        else
-        {
-          throw std::runtime_error(fmt::format(
-            "JS application cannot access private internal CCF table '{}'",
-            table_name));
+          // Private tables should not be accessed at all when executing
+          // governance
+          case TxAccess::GOV_RO:
+          case TxAccess::GOV_RW:
+          {
+            return MapAccessPermissions::ILLEGAL;
+          }
+
+          case TxAccess::APP:
+          {
+            switch (namespace_of_table)
+            {
+              // Private tables in the internal and governance namespaces should
+              // not exist. But if they do, they must be read-only for
```

Comment:
Why should they be read-only and not illegal? We may legitimately have internal private tables (for example for storing inbound requests), and we definitely would not want them readable by application code!

Reply:
As above, I've now changed this behaviour.

I think that's simplified the table a chunk too, so I've also refactored this code to more closely follow from the table in the docs - hopefully the 1-to-1 correspondence between the 2 is easier to track now.